docsTech/agnostic-styling.md

```diff
@@ -0,0 +1,133 @@
+
+const GridQuickFilterTrigger = styled(ToolbarButton, {
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+})(({ ownerState }: { ownerState: OwnerState }) => ({
+  zIndex: 1,
+  opacity: ownerState.expanded ? 0 : 1,
+  pointerEvents: ownerState.expanded ? 'none' : 'auto',
+  transition: vars.transition(['opacity']),
+}));
+
+// usage
+const classes = useUtilityClasses(rootProps)
+<GridQuickFilterTrigger
+  ownerState={{ ...rootProps, expanded: true }}
+  className={classes.trigger}
+/>
+```
+
+#### Proposed syntax
+
+```javascript.jsx
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: {
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+  },
+  expanded: {
+    opacity: 0,
+    pointerEvents: 'none',
+  },
+}));
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (emotion/styled)
+
+```javascript.jsx
+const triggerSlot = {
+  component: styled(ToolbarButton, {
+    name: 'MuiDataGrid',
+    slot: 'toolbarQuickFilterTrigger',
+  })({
+    /* root: */
+    zIndex: 1,
+    opacity: 1,
+    pointerEvents: 'auto',
+    transition: vars.transition(['opacity']),
+
+    /* expanded: */
+    '&.MuiDataGrid-toolbarQuickFilterTrigger--expanded': {
+      opacity: 0,
+      pointerEvents: 'none',
+    },
+  }),
+  classes: {
+    root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+    expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+  },
+}
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Proposed compilation (no emotion)
+
+```javascript.jsx
+import './[currentFilename].autogenerated.css';
+
+const triggerSlot = slot({
+  name: 'MuiDataGrid',
+  slot: 'toolbarQuickFilterTrigger',
+  as: ToolbarButton,
+}, {
+  root: 'MuiDataGrid-toolbarQuickFilterTrigger',
+  expanded: 'MuiDataGrid-toolbarQuickFilterTrigger--expanded',
+}))
+
+// usage
+const trigger = useGridStyleSlot(rootProps, triggerSlot)
+<trigger.component
+  ownerState={rootProps}
+  className={clsx(
+    trigger.classes.root,
+    expanded && trigger.classes.expanded,
+  )}
+/>
+```
+
+#### Downsides
+
+- Right now, the development mode uses the packages source code directly to run. If we require a build step to turn the core grid codebase into emotion/non-emotion builds, the DX for working on the material grid might become worst. Keep in mind that in this scenario, most of the datagrid codebase would be inside the core grid package(s), while only leaf material code would live inside the `@mui/x-data-grid{,pro,premium}` packages.
```

Comment:
> Right now, the development mode uses the packages source code directly to run.

We're using the same Babel plugin to extract the styles both in dev and at the build step, right?

To have emotion/non-emotion builds, we'll likely have 2 Babel plugins with different outputs.
We'll run the code through both Babel plugins at the build step.
In the material data grid's dev mode, we can use the Babel plugin, which outputs an emotion build of the data grid. Would this work?